src/Microsoft.AspNetCore.Routing/Matching/DfaMatcherBuilder.cs

```diff
@@ -500,11 +527,22 @@ internal Candidate CreateCandidate(Endpoint endpoint, int score)
                         slotIndex = _assignments.Count;
                         _assignments.Add(parameterPart.Name, slotIndex);
 
-                        var hasDefaultValue = parameterPart.Default != null || parameterPart.IsCatchAll;
-                        _slots.Add(hasDefaultValue ? new KeyValuePair<string, object>(parameterPart.Name, parameterPart.Default) : default);
+                        if (RequiredValueHelpers.TryGetRequiredValue(routeEndpoint.RoutePattern, parameterPart, out var requiredValue))
+                        {
+                            _slots.Add(new KeyValuePair<string, object>(parameterPart.Name, requiredValue));
+                        }
+                        else
+                        {
+                            var hasDefaultValue = parameterPart.Default != null || parameterPart.IsCatchAll;
+                            _slots.Add(hasDefaultValue ? new KeyValuePair<string, object>(parameterPart.Name, parameterPart.Default) : default);
+                        }
```

Comment:
I think this needs some comments explaining what cases as possible. I had to read this a few times to convince myself that it's right 👍